pyem/metadata.py

```diff
@@ -292,23 +292,80 @@ def cryosparc_2_cs_ctf_parameters(cs, df=None):
     log = logging.getLogger('root')
     if df is None:
         df = pd.DataFrame()
-    if 'ctf/tilt_A' in cs.dtype.names:
-        log.debug("Recovering beam tilt and converting to mrad")
-        df[star.Relion.BEAMTILTX] = np.arcsin(cs['ctf/tilt_A'][:, 0] / cs['ctf/cs_mm'] * 1e-7) * 1e3
-        df[star.Relion.BEAMTILTY] = np.arcsin(cs['ctf/tilt_A'][:, 1] / cs['ctf/cs_mm'] * 1e-7) * 1e3
-    if 'ctf/shift_A' in cs.dtype.names:
-        pass
-    if 'ctf/trefoil_A' in cs.dtype.names:
-        pass
-        # df[star.Relion.ODDZERNIKE] = cs['ctf/trefoil_A']
-    if 'ctf/tetrafoil_A' in cs.dtype.names:
-        pass
-        # df[star.Relion.EVENZERNIKE] = cs['ctf/tetra_A']
+    if 'ctf/accel_kv' in cs.dtype.names:
+        # Calculate electron wavelentgh
+        # https://www.jeol.co.jp/en/words/emterms/search_result.html?keyword=wavelength%20of%20electron
+        E = cs['ctf/accel_kv']  # convert units?
+        wavelength = 1.23e3 / np.sqrt(E*(1 + 9.78e-7 * E))  # ? TODO: check this
+    else:
+        wavelength = None
     if 'ctf/anisomag' in cs.dtype.names:
         df[star.Relion.MAGMAT00] = cs['ctf/anisomag'][:, 0]
         df[star.Relion.MAGMAT01] = cs['ctf/anisomag'][:, 1]
         df[star.Relion.MAGMAT10] = cs['ctf/anisomag'][:, 2]
         df[star.Relion.MAGMAT11] = cs['ctf/anisomag'][:, 3]
+    if 'ctf/tilt_A' in cs.dtype.names:
+        log.debug("Recovering beam tilt and converting to mrad")
+        df[star.Relion.BEAMTILTX] = np.arcsin(cs['ctf/tilt_A'][:, 0] / cs['ctf/cs_mm'] * 1e-7) * 1e3
+        df[star.Relion.BEAMTILTY] = np.arcsin(cs['ctf/tilt_A'][:, 1] / cs['ctf/cs_mm'] * 1e-7) * 1e3
+        df[star.Relion.Z_neg1_3] = df[star.Relion.BEAMTILTX]  # does this also require a conversion factor of 2π * Cs * λ^2
+        df[star.Relion.Z_pos1_3] = df[star.Relion.BEAMTILTY]  # does this also require a conversion factor of 2π * Cs * λ^2
+    if 'ctf/shift_A' in cs.dtype.names:
+        df[star.Relion.Z_neg1_1] = (2 * np.pi) + cs['ctf/shift_A'][:, 0]
+        df[star.Relion.Z_pos1_1]  = (2 * np.pi) + cs['ctf/shift_A'][:, 1]
+    if 'ctf/trefoil_A' in cs.dtype.names and 'ctf/cs_mm' in cs.dtype.names and wavelength is not None:
+        unit_conversion_even = (2*np.pi) * cs['ctf/cs_mm'] * (wavelength**2)
+        df[star.Relion.Z_neg3_3] = unit_conversion_even * cs['ctf/trefoil_A'][:, 0]
+        df[star.Relion.Z_pos3_3] = unit_conversion_even * cs['ctf/trefoil_A'][:, 1]
+    if 'ctf/amp_contrast' in cs.dtype.names:
+        df[star.Relion.Z_0_0] = cs['ctf/amp_contrast'] - np.arccos(phi)  # what is phi? Could it be cs['ctf/phas_shift_rad'] ?
+    if 'ctf/df1_A' in cs.dtype.names and 'ctf/df2_A' in cs.dtype.names and wavelength is not None:
+        df[star.Relion.Z_neg2_2] = np.pi * wavelength * cs['ctf/df1_A'] # defocus asigmatism (X) Z1
+        # df[star.Relion.Z_0_2] = np.pi * wavelength * # average defocus  # how to find this? I have ctf/df_angle_rad ctf/bfactor ctf/scale & ctf/scale_const
```

Comment:
I'm not sure how to calculate the average defocus, could you clarify what should happen here?

Reply:
The average defocus is just `df_avg = (df1_A + df2_A) / 2`. With the normal major/minor axis representation, that quantity, the "defocus deviation" `df_dev = (df1_A  - df2_A) / 2`, and the astigmatism angle are used to actually evaluate the CTF.
 
To describe defocus for the Zernike polynomial model, Relion uses a quadratic form that's related to the usual representation by Eqn 51 of the higher-order aberrations paper.
k^T D k = π*λ*(df_avg + df_dev * cos^2(α) - df_dev * sin^2(α)).

I think the coefficients are simplified terms from this equation, e.g. `Z_0_2 = π * λ * (df_avg)` just like you have already and then if we apply the double-angle formula and then divide out df_dev we get
`Z_neg2_2 = π * λ * cos(2*α) * def_dev`
`Z_2_2 = π * λ * sin(2*α) * def_dev`

Reply:
> The average defocus is just `df_avg = (df1_A + df2_A) / 2`

I had the feeling this was going to be so incredibly obvious it was going to be a little embarrassing 😆 Thanks!

Reply:
Ok, I think I have taken care of the defocus related parameters properly now